checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -61,19 +67,32 @@
 import org.checkerframework.framework.util.dependenttypes.DependentTypesHelper;
 import org.checkerframework.javacutil.AnnotationBuilder;
 import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ElementUtils;
 import org.checkerframework.javacutil.TreeUtils;
 
 /**
  * Implements the introduction rules for the Upper Bound Checker. Works primarily by way of querying
  * the MinLen Checker and comparing the min lengths of arrays to the known values of variables as
  * supplied by the Value Checker.
  */
-public class UpperBoundAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+public class UpperBoundAnnotatedTypeFactory
+        extends GenericAnnotatedTypeFactory<
+                CFValue, UpperBoundStore, UpperBoundTransfer, UpperBoundAnalysis> {
 
     public final AnnotationMirror UNKNOWN, BOTTOM, POLY;
 
     private final IndexMethodIdentifier imf;
 
+    /**
+     * A map that relates an element representing a class to all the receivers of elements of the
+     * class that are dependend on by user-written annotations.
+     */
+    private final HashMap<Element, List<FlowExpressions.Receiver>> dependentReceiversByClass;
```

Comment:
Caches should be cleared between compilation in a method overriding AnnotatedTypeFactory#setRoot.  Otherwise, these caches will be huge!